### PR TITLE
decode_view from pct_string_view

### LINKED ISSUE #534: decode_view from pct_string_view
`decode_view` should only be constructible from `pct_string_view` and the Exception Safety and @throw clauses should match.

### PR DESCRIPTION
**Pull request by alandefreitas**
fix #534

### CODE REVIEW

File: include/boost/url/decode_view.hpp [hunk 1]
```diff
@@ -1001,6 +1005,24 @@ operator<<(
...
+
+    @par Exception Safety
+    Throws nothing.
+
+    @see
+        @ref decode_view.
+*/
+inline
+decode_view
+operator*(pct_string_view s) noexcept
```
Comment (vinniefalco) [inline include/boost/url/decode_view.hpp:1021]: IF you do this then the operator will just be by itself in the docs and not attached to `pct_string_view` where it should be...

File: include/boost/url/pct_string_view.hpp [hunk 2]
```diff
@@ -173,6 +172,36 @@ class pct_string_view final
...
+        Exceptions thrown on invalid input.
+
+        @throw system_error
+        The string contains an invalid percent encoding.
+
+        @param s The string to construct from.
+    */
+    template <class Allocator>
+    pct_string_view(
+        std::basic_string<char, Allocator> const& s)
```
Comment (vinniefalco) [inline include/boost/url/pct_string_view.hpp:199]: something ain't right here... if you are going to add a templated constructor then shouldn't it be anything which is convertible to `urls::string_view`?

File: test/unit/rfc/query_rule.cpp [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include <boost/url/rfc/query_rule.hpp>
 
 #include <boost/url/grammar/range_rule.hpp>
+#include <boost/url/decode_view.hpp>
```
Comment (vinniefalco) [inline test/unit/rfc/query_rule.cpp:14]: these includes are not in the right order

File: include/boost/url/pct_string_view.hpp (hunk 2 above)
Comment (alandefreitas) [reply include/boost/url/pct_string_view.hpp:199]: > shouldn't it be anything which is convertible to urls::string_view

That became ambiguous now. For the same reason, we need the `pct_string_view(char const* s)` just above this one.

File: include/boost/url/decode_view.hpp [hunk 4]
```diff
@@ -180,20 +181,23 @@ class decode_view
         @par Exception Safety
         Exceptions thrown on invalid input.
 
+        @throw std::invalid_argument `s` is
+        not a valid percent-encoded string.
```
Comment (vinniefalco) [inline include/boost/url/decode_view.hpp:185]: Wrong exception

File: include/boost/url/decode_view.hpp [hunk 5]
```diff
@@ -180,20 +181,23 @@ class decode_view
...
-
-        @throw std::invalid_argument `s` is
-        not a valid percent-encoded string.
     */
     BOOST_URL_DECL
     explicit
     decode_view(
-        string_view s,
-        decode_opts opt = {});
+        pct_string_view s,
```
Comment (vinniefalco) [inline include/boost/url/decode_view.hpp:196]: Where's the macro?

File: include/boost/url/detail/config.hpp [hunk 6]
```diff
@@ -110,4 +110,10 @@ using pos_t = size_t;
 #endif
 #endif
 
+#if defined(BOOST_URL_DOCS)
+#define BOOST_URL_PCT_STRING_VIEW string_view
```
Comment (vinniefalco) [inline include/boost/url/detail/config.hpp:114]: I think you have to do this in the Jamfile, not like this

Comment (alandefreitas) [reply include/boost/url/detail/config.hpp]: Good idea. Done.

File: include/boost/url/pct_string_view.hpp (hunk 2 above)
Comment (vinniefalco) [reply include/boost/url/pct_string_view.hpp:200]: why not
```
template<
    class StringLike
    , class = typename std::enable_if<
        std::is_convertible< StringLike, string_view >::value>:: type
>
pct_string_view( StringLike const& );
```

File: include/boost/url/pct_string_view.hpp [hunk 7]
```diff
@@ -152,52 +152,34 @@ class pct_string_view final
...
-
-        @throw system_error
-        The string contains an invalid percent encoding.
+        @tparam StringLike A type convertible to @ref string_view
 
         @param s The string to construct from.
     */
-    template <class Allocator>
+    template<
+        class StringLike
```
Comment (vinniefalco) [inline include/boost/url/pct_string_view.hpp:172]: this actually worked??

Comment (alandefreitas) [reply include/boost/url/pct_string_view.hpp:172]: It worked on GCC, Clang, MSVC, C++11, and C++20 locally. Let's see what CI says. :)

Comment (vinniefalco) [inline include/boost/url/pct_string_view.hpp:172]: Lets call it "String" so users don't get confused... I copied from cppreference which probably wasn't the best.

Comment (alandefreitas) [reply include/boost/url/pct_string_view.hpp]: Done

**Pull request closed without merging**